meta.go

```diff
@@ -229,6 +229,15 @@ func (md *Meta) Set(key string, val interface{}) (err error) {
 	return
 }
 
+// SetKeyVal is for implementing the KeyValSetter defined by base/fill_struct.go
+func (md *Meta) SetKeyVal(key string, val interface{}) (err error) {
```

Comment:
I think this should be a wrapper of `Meta.Set`. What happens when I try to SetKeyVal with a recognized meta field like `Description`?

Reply:
So it really depends upon exactly what the interface's semantics are supposed to be, but I was picturing it the other way. The interface describes only when some user wants to store an arbitrary field in some map[string]-like storage. Otherwise, is the method supposed to completely duplicate the functionality of reflection? That adds a maintenance burden and duplicates code.

For the current usage, in FillStruct, a field like "description" would get handled by the reflection before ending up here, so it's completely arbitrary. I would rather err on the simpler implementation for now, and leave the definition of "SetKeyVal" ambiguous until we have a better idea of its wider usage.

It's also less efficient to use Set instead of SetKeyVal, because every single call has to look through the existing fields, but only barely, so I don't think that actually matters.

Reply:
Hrm... I'm concerned this will lead to unexpected behavior. I think the name for what you're describing would be something like `SetArbitrary(key string, val interface{}) (err error)`. If I were coming to `SetKeyVal` with no prior context, I would assume that I could give this method _any_ key, and it would set it.

I understood `SetKeyVal` to be an opt-in override for the behaviours FillStruct is occupying, a lá [`json.Marshaller`](https://golang.org/pkg/encoding/json/#Marshaler). To me it's an antipattern to define public methods across packages that are solely intended to satisfy single interfaces. Someone at some point will import the dataset package, try to set a description with `SetKeyVal`, and be confused when it doesn't work.

This isn't duplicated code, because Meta does something special when it comes to supporting arbitrary keys. `Set` is implemented without the use of reflection at all, so if FillStruct can hand it's duties off to an interface override, it can bail from using (slow, runtime-driven) reflection earlier, which I suspect would be a performance pickup in this case.

How would you feel about switching to just defining an interface around `Set` in FillStruct, and using an early return? This would require no modifications to the dataset api at all, and remove the need to increase Meta's Public method count

Reply:
Dove into this, and discussed some offline, and I think we're going to go with a rename to `SetArbitrary`.

The issue with switching to `Set`, as it is now, is that `FillStruct` is handling a lot of subtle behavior and unexpected edge cases. For example, all field names get lower-cased to be case-insensitive, and int <-> float coercion has to happen to handle how json unmarshaling works, and these rules need to be passed down recursively to complex field values. `Meta.Set` handles some of this behavior, and `Config.Set` handles some other of this behavior, but neither of these does everything. For example, Meta.Set("citations") ends up being case-sensitive, because `Set` is calling `json.Decode`. The only way to fix all of this would be have these `Set` methods call `FillStruct`, recursively, which implies that they instead should be merged into `FillStruct`. Otherwise, the more we spread this behavior out to other methods, the more likely it is that there's tiny, hard-to-find bugs sitting around our codebases. Better to centralize this handling in `FillStruct` and make `SetArbitrary` a very thin handler for the special-case behavior that `Meta` needs to specifically do.